Add warning for Windows export when rcedit is not configured

NeilKleistGao opened the pull request:
Add warning for Windows export when rcedit is not configured mentioned in #57320.
I think there is no need for checking whether application details are filled or not since the default value(like file version `1.0.0`) will be ignored too if rcedit is not configured. 

*Bugsquad edit:* Fixes #57320.

akien-mga commented:
> I think there is no need for checking whether application details are filled or not since the default value(like file version `1.0.0`) will be ignored too if rcedit is not configured.

`1.0.0` is not a default value, it's a placeholder. So it would still make sense IMO to check if any of those field is not empty (that's what the field is when it's only showing the `1.0.0` placeholder).

Moreover the warning should be shown in the editor UI. A `WARN_PRINT` is better than nothing but it would be better if this was visible directly in the export dialog, like done already for various types of configuration issues.

NeilKleistGao commented:
> > I think there is no need for checking whether application details are filled or not since the default value(like file version `1.0.0`) will be ignored too if rcedit is not configured.
> 
> `1.0.0` is not a default value, it's a placeholder. So it would still make sense IMO to check if any of those field is not empty (that's what the field is when it's only showing the `1.0.0` placeholder).
> 
> Moreover the warning should be shown in the editor UI. A `WARN_PRINT` is better than nothing but it would be better if this was visible directly in the export dialog, like done already for various types of configuration issues.

However, if rcedit is not configured this function will return immediately. Should I still need to check those fields before checking the rcedit path? Thank you for your help and patience. :smile:

bruvzg reviewed (commented):
In addition to the `WARN_PRINT` in the `_rcedit_add_data`, there should be also check for it in the `can_export` method (checks for the valid version/etc. values should go there as well).

See macOS export code for example (warning should only add a text line to the `err`, errors also set `valid = false`).

NeilKleistGao commented:
I added more checks in `can_export` function and now there is a warning message if there is no rcedit configured or wrong version string:![](https://s2.loli.net/2022/01/28/9gMcfa8G6kLYPd3.png)
I also fixed the version placeholders now it will display `1.0.0.0`. :smile:

akien-mga commented on editor/editor_export.cpp:
```diff
@@ -1803,6 +1803,41 @@ bool EditorExportPlatformPC::can_export(const Ref<EditorExportPreset> &p_preset,
+		err += TTR("Invalid icon path: ") + "\n" + icon_path + "\n";
+	}
+
+	String file_verion = p_preset->get("application/file_version");
```
Typo in the identifier (should be `file_version`).

akien-mga commented on editor/editor_export.cpp:
```diff
@@ -1803,6 +1803,41 @@ bool EditorExportPlatformPC::can_export(const Ref<EditorExportPreset> &p_preset,
+
+	String icon_path = ProjectSettings::get_singleton()->globalize_path(p_preset->get("application/icon"));
+	if (!icon_path.is_empty() && !FileAccess::exists(icon_path)) {
+		err += TTR("Invalid icon path: ") + "\n" + icon_path + "\n";
```
```suggestion
		err += TTR("Invalid icon path: ") + icon_path + "\n";
```
Each line gets a `- ` prefix automatically as if it's a new warning, so it's better to keep this one on a single line too.

akien-mga commented on editor/editor_export.cpp:
```diff
@@ -1803,6 +1803,41 @@ bool EditorExportPlatformPC::can_export(const Ref<EditorExportPreset> &p_preset,
+#ifdef WINDOWS_ENABLED
+	String rcedit_path = EditorSettings::get_singleton()->get("export/windows/rcedit");
+	if (rcedit_path.is_empty()) {
+		err += TTR("The path of rcedit is empty. No icon or app information data will be included.") + "\n";
```
```suggestion
		err += TTR("The rcedit tool must be configured in the Editor Settings (Export > Windows > Rcedit) to change the icon or app information data.") + "\n";
```

Calinou commented on editor/editor_export.cpp:
```diff
@@ -1803,6 +1803,41 @@ bool EditorExportPlatformPC::can_export(const Ref<EditorExportPreset> &p_preset,
 	valid = dvalid || rvalid;
 	r_missing_templates = !valid;
 
+#ifdef WINDOWS_ENABLED
```
rcedit can be used under Linux and macOS with wine, so this code should not be specific to Windows editor binaries.

NeilKleistGao replied:
I replaced macro with `get_os_name` and I hope this one can work.

EzraT commented:
Thanks!

Calinou commented on editor/editor_export.cpp:
```diff
@@ -1803,6 +1803,40 @@ bool EditorExportPlatformPC::can_export(const Ref<EditorExportPreset> &p_preset,
 	valid = dvalid || rvalid;
 	r_missing_templates = !valid;
 
+	if (get_os_name() == "Windows") {
```
What I meant with my previous review comment is that this code should be run on macOS and Linux too. However, on all platforms, this code should only run on Windows *export presets*, not on macOS or Linux export presets.

NeilKleistGao replied:
> What I meant with my previous review comment is that this code should be run on macOS and Linux too, as you can export Godot projects for Windows from other platforms. (Apologies if this wasn't clear before.)
> 
> However, on all platforms, this code should only run on Windows _export presets_, not on macOS or Linux export presets.

Thanks for your review and I think it was clear enough. I also believe that this code can work as expected. However, I do need to move it from `EditorExportPlatformPC` to `EditorExportPlatformWindows` since it isn't needed on macOS or Linux.

akien-mga commented on platform/windows/export/export_plugin.cpp:
```diff
@@ -89,6 +89,7 @@ void EditorExportPlatformWindows::_rcedit_add_data(const Ref<EditorExportPreset>
 	String rcedit_path = EditorSettings::get_singleton()->get("export/windows/rcedit");
 
 	if (rcedit_path.is_empty()) {
+		WARN_PRINT("The path of rcedit is empty. No icon or app information data will be included.");
```
```suggestion
		WARN_PRINT("The rcedit tool is not configured in the Editor Settings (Export > Windows > Rcedit). No custom icon or app information data will be embedded in the exported executable.");
```

akien-mga commented on platform/windows/export/export_plugin.cpp:
```diff
@@ -327,3 +328,46 @@ Error EditorExportPlatformWindows::_code_sign(const Ref<EditorExportPreset> &p_p
+
+	String icon_path = ProjectSettings::get_singleton()->globalize_path(p_preset->get("application/icon"));
+	if (!icon_path.is_empty() && !FileAccess::exists(icon_path)) {
+		err += TTR("Invalid icon path: ") + icon_path + "\n";
```
Trailing whitespace is difficult to translate, it's better to keep it out:
```suggestion
		err += TTR("Invalid icon path:") + " " + icon_path + "\n";
```

akien-mga commented on platform/windows/export/export_plugin.cpp:
```diff
@@ -327,3 +328,46 @@ Error EditorExportPlatformWindows::_code_sign(const Ref<EditorExportPreset> &p_p
+		if (version_array.size() != 4 || !version_array[0].is_valid_int() ||
+				!version_array[1].is_valid_int() || !version_array[2].is_valid_int() ||
+				!version_array[3].is_valid_int() || file_version.find("-") > -1) {
+			err += TTR("Invalid file version: ") + file_version + "\n";
```
```suggestion
			err += TTR("Invalid file version:") + " " + file_version + "\n";
```

akien-mga commented on platform/windows/export/export_plugin.cpp:
```diff
@@ -327,3 +328,46 @@ Error EditorExportPlatformWindows::_code_sign(const Ref<EditorExportPreset> &p_p
+		if (version_array.size() != 4 || !version_array[0].is_valid_int() ||
+				!version_array[1].is_valid_int() || !version_array[2].is_valid_int() ||
+				!version_array[3].is_valid_int() || product_version.find("-") > -1) {
+			err += TTR("Invalid product version: ") + product_version + "\n";
```
```suggestion
			err += TTR("Invalid product version:") + " " + product_version + "\n";
```

akien-mga commented:
Please squash the commits into one once this is ready, following the instructions in [PR workflow](https://docs.godotengine.org/en/latest/community/contributing/pr_workflow.html#the-interactive-rebase).

Commit:
Add warning for Windows export when rcedit is not configured

akien-mga commented:
Thanks! And congrats for your first merged Godot contribution :tada:

NeilKleistGao commented:
> Thanks! And congrats for your first merged Godot contribution 🎉

Thank you for your patience and help! 😃

akien-mga commented:
Cherry-picked for 3.5.
